src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -67,6 +67,13 @@ internal partial class Http2Connection : IHttp2StreamLifetimeHandler, IHttpHeade
         private int _gracefulCloseInitiator;
         private int _isClosed;
 
+        private readonly Http2Stream[] _streamPool;
```

Comment:
Why not use a Queue or Stack so you don't have to do your own bookkeeping? It could even be a concurrent one which have better locking.

Reply:
[`ConcurrentQueueSegment<T>`](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Collections/Concurrent/ConcurrentQueueSegment.cs) is the ideal one, removing everything from of it that's not needed for `TryDequeue` and `TryEnqueue`. 

Size must be a power of two; however is ideal for object pooling.

Alas it never was turned into a [BoundedConcurrentQueue<T>](https://github.com/dotnet/corefx/issues/24365) api